lib/transfer-memory-allocator.js

```diff
@@ -0,0 +1,544 @@
+    }
+
+    /**
+     * Get the end index from the buffer pool (highest index/memory
+     * location usable for this Transfer Memory Block)
+     * @returns {Integer} End index of the allocated buffer 
+     * (memory block) in the buffer pool
+     */
+    get endIndex() {
+        return this.startIndex + this.size - 1;
+    }
+
+    /**
+     * Get the buffer representing the memory block from the pooled
+     * buffer allocated for this Transfer Memory Block.
+     * Modifying this memory block (buffer slice) will modify the memory 
+     * in the original Buffer because the allocated memory of this memory block 
+     * and the original buffer pool overlap.
+     */
+    get buffer() {
+        return this[INTERNAL].referencedMemoryFromBufferPool;
+    }
+
+    /**
+     * Set the buffer to use for memory access. 
+     * Used by Transfer Memory Block allocator
+     * @param {Buffer} value Buffer to use for memory access 
+     */
+    set buffer(value) {
+        return this[INTERNAL].referencedMemoryFromBufferPool = value;
+    }
+}
+
+// ------------------------------------------------------------------------------
+const PRIVATE = Symbol("private");
+/**
+ * Keeps track of memory usable for transfers, and allocates available 
+ * memory as needed/when requested by transfers
+ */
+class TransferMemoryBuffer {
+    /**
+     * Constructs a transfer memory buffer
+     * @param {*} suggestedSize Suggested size for the memory reserved 
+     * for buffering transfer data, in bytes. The memory allocator will
+     * try to allocate memory with size `suggestedSize`, but if not 
+     * enough memory is available, 80% of available memory will be 
+     * allocated instead
+     */
+    constructor(suggestedSize) {
+        this[PRIVATE] = {};
+
+        let bufferSize;
+        if (suggestedSize) {
+            console.log(`Suggested Transfer Memory Buffer pool size is ${suggestedSize} bytes`);
+            bufferSize = parseInt(suggestedSize, 10);
+        } else {
+            console.log(`Using default size for Transfer Memory Buffer: ${suggestedSize} bytes`);
+            bufferSize = DEFAULT_TRANSFER_BUFFER_POOL_SIZE;
+        }
+        if (suggestedSize <= 0) {
+            throw new Error("Transfer Memory Buffer total suggested size must be larger than 0");
+        }
+
+        const availableFreeMemory = os.freemem();
+        const highMemoryMark = Math.floor(availableFreeMemory * MAX_PERCENTAGE_ALLOWED_MEMORY_USE);
+        if (bufferSize > highMemoryMark) {
+            // use 80% of available memory
+            this[PRIVATE].totalSize = Math.min(highMemoryMark, MAX_BUFFER_SIZE);
+            console.log(`Not enough memory available (${availableFreeMemory} bytes available, ${bufferSize} bytes requested) to use suggested size for Transfer Memory Buffer pool size. Using ${MAX_PERCENTAGE_ALLOWED_MEMORY_USE * 100}% of available memory instead: ${this[PRIVATE].totalSize} bytes as Transfer Memory Buffer pool size`);
```

Comment:
can you mock the max buffer size or high mem mark to make a test that hits this case?

Reply:
I tried - but the mocking for the `os` module seems to be flaky.